File: backend/apps/users/validators.py
"""
Plain-function validators for user input. Kept separate from serializers so
they're easy to unit test and reuse (e.g. re-validating email format if we
ever add "update profile").
"""

import re

EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


def validate_email_format(email):
    if not email or not EMAIL_RE.match(email):
        return "Enter a valid email address."
    return None


def validate_password_strength(password):
    if not password or len(password) < 8:
        return "Password must be at least 8 characters long."
    if not re.search(r"[A-Za-z]", password) or not re.search(r"[0-9]", password):
        return "Password must contain at least one letter and one number."
    return None


def validate_name(name):
    if not name or not name.strip():
        return "Name is required."
    return None

File: frontend/src/utils/validators.js
/**
 * Frontend form validation. These mirror (but do not replace) backend
 * validation - the backend remains the source of truth, this just gives
 * the user faster, friendlier feedback before a round trip.
 */

export function validateEmail(value) {
  if (!value) return "Email is required.";
  if (!/^[^@\s]+@[^@\s]+\.[^@\s]+$/.test(value)) return "Enter a valid email address.";
  return "";
}

export function validatePassword(value) {
  if (!value) return "Password is required.";
  if (value.length < 8) return "Password must be at least 8 characters.";
  if (!/[a-zA-Z]/.test(value) || !/[0-9]/.test(value)) {
    return "Password must contain at least one letter and one number.";
  }
  return "";
}

export function validateConfirmPassword(password, confirm) {
  if (!confirm) return "Please confirm your password.";
  if (password !== confirm) return "Passwords do not match.";
  return "";
}

export function validateName(value) {
  if (!value || !value.trim()) return "Name is required.";
  return "";
}

export function validateTaskTitle(value) {
  if (!value || !value.trim()) return "Title is required.";
  if (value.trim().length > 200) return "Title must be under 200 characters.";
  return "";
}
